src/main/scala/org/opennetworkinsight/dns/DNSSuspiciousConnectsAnalysis.scala

```diff
@@ -0,0 +1,93 @@
+package org.opennetworkinsight.dns
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.opennetworkinsight.SuspiciousConnectsArgumentParser.SuspiciousConnectsConfig
+import org.opennetworkinsight.dns.DNSSchema._
+import org.opennetworkinsight.dns.model.DNSSuspiciousConnectsModel
+import org.opennetworkinsight.dns.sideinformation.DNSSideInformation
+import org.slf4j.Logger
+
+import org.opennetworkinsight.dns.model.DNSSuspiciousConnectsModel.ModelSchema
+import org.opennetworkinsight.dns.sideinformation.DNSSideInformation.SideInfoSchema
+
+/**
+  * The suspicious connections analysis of DNS log data develops a probabilistic model the DNS queries
+  * made by each client IP and flags
+  */
+
+object DNSSuspiciousConnectsAnalysis {
+
+  val inSchema = StructType(List(TimestampField, UnixTimestampField, FrameLengthField, ClientIPField,
+      QueryNameField, QueryClassField, QueryTypeField, QueryResponseCodeField))
+
+  val inColumns = inSchema.fieldNames.map(col)
+
+
+  assert(ModelSchema.fields.forall(inSchema.fields.contains(_)))
+
+  val OutSchema = StructType(
+    List(TimestampField,
+      UnixTimestampField,
+      FrameLengthField,
+      ClientIPField,
+      QueryNameField,
+      QueryClassField,
+      QueryTypeField,
+      QueryResponseCodeField,
+      DomainField,
+      SubdomainField,
+      SubdomainLengthField,
+      NumPeriodsField,
+      SubdomainEntropyField,
+      TopDomainField,
+      WordField,
+      ScoreField))
+
+  val OutColumns = OutSchema.fieldNames.map(col)
+
+
+  /**
+    * Run suspicious connections analysis on DNS log data.
+    *
+    * @param config Object encapsulating runtime parameters and CLI options.
+    * @param sparkContext
+    * @param sqlContext
+    * @param logger
+    */
+  def run(config: SuspiciousConnectsConfig, sparkContext: SparkContext, sqlContext: SQLContext, logger: Logger) = {
+    logger.info("Starting DNS suspicious connects analysis.")
+
+    val topicCount = 20
+
+    logger.info("Loading data")
+
+    val rawDataDF = sqlContext.read.parquet(config.inputPath)
+      .filter(Timestamp + " is not null and " + UnixTimestamp + " is not null")
+      .select(inColumns:_*)
+
+    logger.info("Training the model")
+
+    val model =
+      DNSSuspiciousConnectsModel.trainNewModel(sparkContext, sqlContext, logger, config, rawDataDF, topicCount)
+
+    logger.info("Scoring")
+    val scoredDF = model.score(sparkContext, sqlContext, rawDataDF)
+
+
+    val filteredDF = scoredDF.filter(Score + " <= " + config.threshold)
+    val mostSusipiciousDF: DataFrame = filteredDF.orderBy(Score).limit(config.maxResults)
+
+    // add the "side information" expected by the OA layer
+    val sideInformationGenerator = new DNSSideInformation(model)
```

Comment:
there won't be any changes in this PR, but we should get the ball rolling for future discussion

in theory, I believe that each piece of code should have a single responsibility 
in practice, I recognize that sometimes it's way cheaper to spin up one batch job that does it all than to do a bunch of one-off jobs